### ISSUE TITLE
kube-system pods toleration "node-role.kubernetes.io/master" should not have any value

### ISSUE THREAD

**Issue by bclau** [Azure/aks-engine#766]
**Is this a request for help?**: No

---

**Is this an ISSUE or FEATURE REQUEST?** (choose one): ISSUE

---

**What version of aks-engine?**: v0.31.3

---

**Kubernetes version**: v1.14.0-alpha.1

**What happened**:

Typically, the master nodes should be tainted as follows: 

    kubectl taint nodes k8s-master-38923293-0 node-role.kubernetes.io/master=:NoSchedule

As you can see, the taint doesn't have a value, and having no value is considered standard practice [1][2]. This is further backed by the fact that the ``coredns`` pods has the the toleration ``node-role.kubernetes.io/master:NoSchedule``.

[1] https://github.com/kubernetes-sigs/kubespray/issues/2108
[2] https://kubernetes.io/docs/setup/independent/create-cluster-kubeadm/

But, looking at the tolerations set to pods in the kube-system namespace, it can be seen that only the ``coredns`` pod has the ``node-role.kubernetes.io/master:NoSchedule`` toleration, while the pods ``azure-cni-networkmonitor, azure-ip-masq-agent, kube-proxy`` have the toleration ``node-role.kubernetes.io/master=true:NoSchedule``:

    kubectl get pods -n kube-system -o json | jq -r '.items[] | select(.status.phase == "Running" and ([ .spec.tolerations[] | select(.key == "node-role.kubernetes.io/master") ] | length ) == 1 ) | .metadata.namespace + "/" + .metadata.name + ": " + (.spec.tolerations[] | select(.key == "node-role.kubernetes.io/master") | .key + ":" + .effect + " " + .operator + " " + .value)'
    kube-system/azure-cni-networkmonitor-xwmvh: node-role.kubernetes.io/master:NoSchedule Equal true
    kube-system/azure-ip-masq-agent-5tkbz: node-role.kubernetes.io/master:NoSchedule Equal true
    kube-system/coredns-d5b7bc49d-58tn5: node-role.kubernetes.io/master:NoSchedule
    kube-system/kube-proxy-k9lzg: node-role.kubernetes.io/master:NoSchedule Equal true

The 2 taints ``node-role.kubernetes.io/master:NoSchedule`` and ``node-role.kubernetes.io/master=true:NoSchedule`` are **not** the same and they **cannot exist** at the same time. Because of this, when applying one of the taints, it can lead to unstable Kubernetes clusters, especially after reboot, since either 1 or 3 out of the 4 pods in questions will not be able to re-enter a Running state.

**What you expected to happen**:

All pods should have the same tolerance, which is ``node-role.kubernetes.io/master:NoSchedule`` (standard tolerance), and not ``node-role.kubernetes.io/master=true:NoSchedule``.

**How to reproduce it** (as minimally and precisely as possible):

Create a new cluster. Execute the command:

    kubectl get pods -n kube-system -o json | jq -r '.items[] | select(.status.phase == "Running" and ([ .spec.tolerations[] | select(.key == "node-role.kubernetes.io/master") ] | length ) == 1 ) | .metadata.namespace + "/" + .metadata.name + ": " + (.spec.tolerations[] | select(.key == "node-role.kubernetes.io/master") | .key + ":" + .effect + " " + .operator + " " + .value)'

**Anything else we need to know**:

**Comment by jackfrancis** [comment]
Hi @bclau, thanks for reporting and providing good detail.

You point out that a node can't have both of these taints at the same time. Are workloads able to have both of these *tolerations* at the same time? That would make the upgrade path workable for pre-existing clusters, because as soon as we make the change to the taint, we will be breaking master-schedulable pods (e.g., kube-proxy), which are as currently defined only able to tolerate "master=true" taints.

In other words, can we define `kube-proxy` (and other addons) thusly:

```
      - key: node-role.kubernetes.io/master
        operator: Equal
        value: "true"
        effect: NoSchedule
      - key: node-role.kubernetes.io/master
        operator: Equal
        effect: NoSchedule
```

Can you define a workload with both of those tolerations, so that those workloads will be able to be scheduled on a master which has either one of them?

**Comment by bclau** [comment]
Indeed, pods can be created with *both* tolerations: https://paste.ubuntu.com/p/WYbgPnmrCx/
